common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java

```diff
@@ -57,18 +63,121 @@ public Context getContext() {
         return mContext;
     }
 
+    /**
+     *
+     * @return handler completion callback
+     */
+    IChallengeCompletionCallback getCompletionCallback() {
+        return mCompletionCallback;
+    }
+
     /**
      * Constructor for the OAuth2 basic web view client.
-     * @param context app Context
-     * @param request Authorization request
+     *
+     * @param context  app Context
+     * @param request  Authorization request
+     * @param callback Challenge completion callback
      */
-    OAuth2WebViewClient(@NonNull final Context context, @NonNull final AuthorizationRequest request) {
+    OAuth2WebViewClient(@NonNull final Context context,
+                        @NonNull final AuthorizationRequest request,
+                        @NonNull final IChallengeCompletionCallback callback) {
         //the validation of redirect url and authorization request should be in upper level before launching the webview.
-        if (null == context || null == request) {
-            throw new IllegalArgumentException("Null parameter to initialize OAuth2WebViewClient.");
-        }
-
         mContext = context;
         mRequest = request;
+        mCompletionCallback = callback;
+    }
+
+    @Override
+    public void onReceivedHttpAuthRequest(WebView view, final HttpAuthHandler handler,
+                                          String host, String realm) {
+        // Create a dialog to ask for credentials and post it to the handler.
+        Logger.info(TAG, "Receive the http auth request. Start the dialog to ask for creds. ");
+        Logger.infoPII(TAG, "Host:" + host);
+
+        //TODO TelemetryEvent.setNTLM(true); after the Telemetry is finished in common.
+        final NtlmChallengeHandler ntlmChallengeHandler

```

Comment:
The order of events and object responsibilities feel a little mixed here.

Currently:
The `NtlmChallengeHandler` is created.
The `NtlmChallengeHandler` creates the `NtlmChallenge` (So it's both a handler _and_ a factory)
The `NtlmChallengeHandler` then processes the `NtlmChallenge` that it created. (Then it does some processing)

I suggest we split these actions and responsibilities such that creating the challenge, creating the handler, and doing the processing are 3 discreet, testable, operations - to spitball some pseudocode...

```java
// We have a Factory capable of producing different kinds of challenges
NtlmChallenge ntlmChallenge = new ChallengeFactory().newNtlmChallenge(view, handler, host, realm);

// We init our ChallengeHandler, could conceivably respond to different types of challenges
IChallengeHandler<NtlmChallenge, Void> challengeHandler = new NtlmChallengeHandler();

// We use the ChallengeHandler
challengeHandler.processChallenge(ntlmChallenge);
```

By separating the creation of the `NtlmChallenge` from the `ChallengeHandler` we create a new testable surface in the factory. We can also make the `ChallengeHandler` API more generic by getting rid of the coupling between the Handler and the type of Challenge that's constructed.